sdk/attestation/azure-security-attestation/inc/azure/attestation/attestation_client.hpp

```diff
@@ -0,0 +1,324 @@
+
+    /**
+     * @brief Construct a new Attestation Client object
+     *
+     * @param endpoint The URL address where the client will send the requests to.
+     * @param credential OPTIONAL The authentication method to use (required for TPM attestation).
+     * @param options The options to customize the client behavior.
+     */
+    explicit AttestationClient(
+        std::string const& endpoint,
+        std::shared_ptr<Core::Credentials::TokenCredential const> credential,
+        AttestationClientOptions options = AttestationClientOptions());
+
+    explicit AttestationClient(
+        std::string const& endpoint,
+        AttestationClientOptions options = AttestationClientOptions())
+        : AttestationClient(endpoint, nullptr, options)
+    {
+    }
+
+    /**
+     * @brief Returns the API version the client was configured with.
+     *
+     * @returns The API version used when communicating with the attestation service.
+     */
+    std::string const& ClientVersion() const { return m_apiVersion; }
+
+    /**
+     * @brief Construct a new Key Client object from another key client.
+     *
+     * @param attestationClient An existing attestation client.
+     */
+    explicit AttestationClient(AttestationClient const& attestationClient)
+        : m_endpoint(attestationClient.m_endpoint), m_apiVersion(attestationClient.m_apiVersion),
+          m_pipeline(attestationClient.m_pipeline),
+          m_tokenValidationOptions(attestationClient.m_tokenValidationOptions){};
+
+    /**
+     * Retrieves metadata about the attestation signing keys in use by the attestation service.
+     *
+     * Retrieve the OpenID metadata for this attestation service instance..
+     *
+     * @return an {@link AttestationOpenIdMetadata} containing metadata about the specified service
+     * instance.
+     */
+    Response<Models::AttestationOpenIdMetadata> GetOpenIdMetadata(
+        Azure::Core::Context const& context = Azure::Core::Context::ApplicationContext) const;
+
+    /**
+     * @brief Retrieve the attestation signing certificates for this attestation instance.
+     *
+     * @returns Attestation Metadata.
+     */
+    Response<Models::AttestationSigningCertificateResult> GetAttestationSigningCertificates(
+        Azure::Core::Context const& context = Azure::Core::Context::ApplicationContext) const;
+
+    /**
+     * @brief Attest an SGX enclave, returning an attestation token representing the result
+     * of the attestation operation.
+     *
+     * @param sgxQuoteToAttest - SGX Quote to be validated by the attestation service.
+     * @param options - Options to the attestation request (runtime data, inittime data, etc).
+     * @param context - Context for the operation.
+     *
+     * @returns Response<{@link AttestationToken}<{@link AttestationResult}>> - The result of the
+     * attestation operation
+     */
+    Response<AttestationToken<Models::AttestationResult>> AttestSgxEnclave(
```

Comment:
This is an interesting approach and design (returning a `AttestationToken<T>` templated type which isn't a "model"). I am wondering how other language SDKs, like .NET expose the return type, and if it is similar to what we are doing here.

Reply:
I would actually prefer to return an `AttestationResponse` which derives from `Response`, however `Response` is final :(. That's what I did in the other languages.